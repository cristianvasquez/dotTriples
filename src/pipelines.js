import { createReadStream, stat} from 'fs'
import { resolve } from 'path'
import rdf from 'rdf-ext'
import { Transform } from 'stream'
import { createMarkdownParser } from './markdownParser.js'
import ns from './namespaces.js'
import { ParseDotTriples } from './transforms/parseDotTriples.js'
import { ParseMarkdown } from './transforms/parseMarkdown.js'
import { ProduceQuads } from './transforms/produceQuads.js'

// Expects markdown files, and produces triples
function createMarkdownPipeline ({ basePath, uriResolver, quadProducers }, destStream) {






  const markdownParser = createMarkdownParser()
  const transform = new Transform({
    objectMode: true, transform (path, enc, done) {

      const filePath = resolve(basePath, path)

      stat(filePath, (err, stats) => {
        if (err) {
          console.error(err);
        } else {
          const fileStream = createReadStream(filePath).
            pipe(new ParseMarkdown({ markdownParser }, { path }, {})).
            pipe(new ParseDotTriples()).
            pipe(new ProduceQuads({ uriResolver, quadProducers }, {}))

          fileStream.pipe(destStream, { end: false })
          fileStream.on('new', (item) => {
            console.log('item', item)
          })
          fileStream.on('error', done)
          fileStream.on('end', ()=>{
            const fileUri = uriResolver.getUriFromPath(path)
            const name = uriResolver.getNameFromPath(path)
            const {size, atime,mtime, ctime} = stats
            destStream.write(rdf.quad(fileUri,ns.dot.path,rdf.literal(path)))
            destStream.write(rdf.quad(fileUri,ns.dot.name,rdf.literal(name)))
            destStream.write(rdf.quad(fileUri,ns.dot.size,rdf.literal(size, ns.xsd.integer)))
            destStream.write(rdf.quad(fileUri,ns.dot.atime,rdf.literal(atime.toISOString(), ns.xsd.dateTime)))
            destStream.write(rdf.quad(fileUri,ns.dot.mtime,rdf.literal(mtime.toISOString(), ns.xsd.dateTime)))
            destStream.write(rdf.quad(fileUri,ns.dot.ctime,rdf.literal(ctime.toISOString(), ns.xsd.dateTime)))
            done()
          } )
        }
      });
    },
  })

  transform.on('error', () => {
    destStream.destroy()
  }).on('finish', () => {
    destStream.end()
  }).on('end', () => {
    destStream.end()
  })

  return transform
}

export { createMarkdownPipeline }
