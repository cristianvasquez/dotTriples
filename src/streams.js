import { PassThrough } from 'stream'

function printJsons () {
  return createOutputStream({
    forEach: (chunk => {
      console.log(JSON.stringify(chunk, null, 2))
    }),
  })
}

function printDatasets () {
  return createOutputStream({
    forEach: (chunk => {
      console.log(chunk.toString())
    }),
  })
}

function printDatasetCounts () {
  return new PassThrough({
    objectMode: true, write (dataset, encoding, callback) {
      this.count = this.count ? this.count + dataset.size : dataset.size
      console.log(`${this.count} quads`)
      this.push(dataset)
      callback()
    },
  })
}

function createOutputStream ({ forEach } = { forEach: x => x }) {
  return new PassThrough({
    objectMode: true, write (object, encoding, callback) {
      forEach(object)
      this.push(object)
      callback()
    },
  })
}

export {
  printJsons,
  printDatasets,
  printDatasetCounts,
  createOutputStream,
}
