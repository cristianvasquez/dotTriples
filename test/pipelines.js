import { expect } from 'expect'
import toMatchSnapshot from 'expect-mocha-snapshot'
import { resolve } from 'path'
import rdf from 'rdf-ext'
import { createTriplifier } from 'rdf-from-markdown'
import { PassThrough } from 'stream'
import ns from '../src/namespaces.js'

import { createMarkdownPipeline } from '../src/pipelines.js'
import { collect } from './support/streams.js'

expect.extend({ toMatchSnapshot })

const basePath = 'test/markdown'

// Config

async function createContext () {

  const triplifier = await createTriplifier(basePath, {
    mappers: {
      'is a': ns.rdf.type, 'are': ns.rdf.type, 'foaf:knows': ns.foaf.knows,
    }, baseNamespace: ns.ex,
  })

  return {
    triplifier, basePath: resolve(basePath),
  }
}

describe('[MarkdownPipeline]', async function () {
  const context = await createContext()
  for (const fileName of context.index.files.filter(x => x.endsWith('.md'))) {

    it(`"${fileName}"`, async function () {

      const outputStream = new PassThrough({
        objectMode: true, write (object, encoding, callback) {
          this.push(object)
          callback()
        },
      })

      // Small version that does not produce dates
      function produceStats ({ fileUri, path, name, stats }) {
        const { size } = stats
        return [
          rdf.quad(fileUri, ns.dot.path, rdf.literal(path), fileUri),
          rdf.quad(fileUri, ns.dot.size, rdf.literal(size, ns.xsd.integer),
            fileUri)]
      }

      const inputStream = createMarkdownPipeline({ ...context, produceStats },
        { outputStream })
      inputStream.write(fileName)
      inputStream.end()

      const [dataset] = await collect(outputStream)

      expect(dataset.toString()).toMatchSnapshot(this)
    })
  }

})
