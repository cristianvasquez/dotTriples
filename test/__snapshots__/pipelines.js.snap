// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[MarkdownPipeline] "./Alice.md" 1`] = `
"<http://example.org/Alice.md> <http://example.org/linked-to> <http://example.org/not-found-in-vault> <http://example.org/Alice.md> .
<http://example.org/Alice.md> <http://dottriples.org/path> \\"./Alice.md\\" .
<http://example.org/Alice.md> <http://dottriples.org/name> \\"Alice\\" .
<http://example.org/Alice.md> <http://dottriples.org/size> \\"26\\"^^<http://www.w3.org/2001/XMLSchema#integer> .
<http://example.org/Alice.md> <http://dottriples.org/atime> \\"2022-09-25T13:46:50.310Z\\"^^<http://www.w3.org/2001/XMLSchema#dateTime> .
<http://example.org/Alice.md> <http://dottriples.org/mtime> \\"2022-09-02T02:12:08.644Z\\"^^<http://www.w3.org/2001/XMLSchema#dateTime> .
<http://example.org/Alice.md> <http://dottriples.org/ctime> \\"2022-09-02T02:12:08.644Z\\"^^<http://www.w3.org/2001/XMLSchema#dateTime> .
"
`;

exports[`[MarkdownPipeline] "./Person.md" 1`] = `
"<http://example.org/Person.md> <http://example.org/has-url> <http://web1.org> <http://example.org/Person.md> .
<http://example.org/Person.md> <http://example.org/has-url> <http://web2.org> <http://example.org/Person.md> .
<http://example.org/Person.md> <http://dottriples.org/path> \\"./Person.md\\" .
<http://example.org/Person.md> <http://dottriples.org/name> \\"Person\\" .
<http://example.org/Person.md> <http://dottriples.org/size> \\"58\\"^^<http://www.w3.org/2001/XMLSchema#integer> .
<http://example.org/Person.md> <http://dottriples.org/atime> \\"2022-09-25T17:25:42.067Z\\"^^<http://www.w3.org/2001/XMLSchema#dateTime> .
<http://example.org/Person.md> <http://dottriples.org/mtime> \\"2022-09-25T17:25:42.059Z\\"^^<http://www.w3.org/2001/XMLSchema#dateTime> .
<http://example.org/Person.md> <http://dottriples.org/ctime> \\"2022-09-25T17:25:42.059Z\\"^^<http://www.w3.org/2001/XMLSchema#dateTime> .
"
`;

exports[`[MarkdownPipeline] "./Yaml.md" 1`] = `
"<http://example.org/Yaml.md> <http://example.org/date-updated> \\"2021-08-05T17:10:34+02:00\\" <http://example.org/Yaml.md> .
<http://example.org/Yaml.md> <http://example.org/tags> \\"cats\\" <http://example.org/Yaml.md> .
<http://example.org/Yaml.md> <http://example.org/tags> \\"dogs\\" <http://example.org/Yaml.md> .
<http://example.org/Yaml.md> <http://example.org/one> \\"Two\\" <http://example.org/Yaml.md> .
<http://example.org/Yaml.md> <http://dottriples.org/path> \\"./Yaml.md\\" .
<http://example.org/Yaml.md> <http://dottriples.org/name> \\"Yaml\\" .
<http://example.org/Yaml.md> <http://dottriples.org/size> \\"110\\"^^<http://www.w3.org/2001/XMLSchema#integer> .
<http://example.org/Yaml.md> <http://dottriples.org/atime> \\"2022-09-25T13:46:50.358Z\\"^^<http://www.w3.org/2001/XMLSchema#dateTime> .
<http://example.org/Yaml.md> <http://dottriples.org/mtime> \\"2022-09-02T02:09:27.532Z\\"^^<http://www.w3.org/2001/XMLSchema#dateTime> .
<http://example.org/Yaml.md> <http://dottriples.org/ctime> \\"2022-09-02T02:09:27.532Z\\"^^<http://www.w3.org/2001/XMLSchema#dateTime> .
"
`;

exports[`[MarkdownPipeline] "./bob/Bob.md" 1`] = `
"<http://example.org/bob/Bob.md> <http://example.org/age> \\"44\\" <http://example.org/bob/Bob.md> .
<http://example.org/bob/Bob.md> <http://example.org/website> <http://bob.org> <http://example.org/bob/Bob.md> .
<http://example.org/bob/Bob.md> <http://xmlns.com/foaf/0.1/knows> <http://example.org/Alice.md> <http://example.org/bob/Bob.md> .
<http://example.org/bob/Bob.md> <http://example.org/linked-to> <http://example.org/Alice.md> <http://example.org/bob/Bob.md> .
<http://example.org/bob/Bob.md> <http://example.org/linked-to> <http://example.org/bob/BobHouse.md> <http://example.org/bob/Bob.md> .
<http://example.org/bob/Bob.md> <http://example.org/has-relative> <http://example.org/folder/Test.md> <http://example.org/bob/Bob.md> .
<http://example.org/bob/Bob.md> <http://example.org/has-image> <http://example.org/folder/img.png> <http://example.org/bob/Bob.md> .
<http://bob.org> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> \\"WeBsiTe\\" <http://example.org/bob/Bob.md> .
<http://example.org/bob/Bob.md> <http://dottriples.org/path> \\"./bob/Bob.md\\" .
<http://example.org/bob/Bob.md> <http://dottriples.org/name> \\"Bob\\" .
<http://example.org/bob/Bob.md> <http://dottriples.org/size> \\"331\\"^^<http://www.w3.org/2001/XMLSchema#integer> .
<http://example.org/bob/Bob.md> <http://dottriples.org/atime> \\"2022-09-25T13:46:50.326Z\\"^^<http://www.w3.org/2001/XMLSchema#dateTime> .
<http://example.org/bob/Bob.md> <http://dottriples.org/mtime> \\"2022-09-02T05:14:07.241Z\\"^^<http://www.w3.org/2001/XMLSchema#dateTime> .
<http://example.org/bob/Bob.md> <http://dottriples.org/ctime> \\"2022-09-02T05:14:07.241Z\\"^^<http://www.w3.org/2001/XMLSchema#dateTime> .
"
`;

exports[`[MarkdownPipeline] "./bob/BobHouse.md" 1`] = `
"<http://example.org/bob/BobHouse.md> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> \\"House\\" <http://example.org/bob/BobHouse.md> .
<http://example.org/bob/BobHouse.md> <http://dottriples.org/path> \\"./bob/BobHouse.md\\" .
<http://example.org/bob/BobHouse.md> <http://dottriples.org/name> \\"BobHouse\\" .
<http://example.org/bob/BobHouse.md> <http://dottriples.org/size> \\"14\\"^^<http://www.w3.org/2001/XMLSchema#integer> .
<http://example.org/bob/BobHouse.md> <http://dottriples.org/atime> \\"2022-09-25T13:46:50.342Z\\"^^<http://www.w3.org/2001/XMLSchema#dateTime> .
<http://example.org/bob/BobHouse.md> <http://dottriples.org/mtime> \\"2022-09-02T03:38:48.794Z\\"^^<http://www.w3.org/2001/XMLSchema#dateTime> .
<http://example.org/bob/BobHouse.md> <http://dottriples.org/ctime> \\"2022-09-02T03:38:48.794Z\\"^^<http://www.w3.org/2001/XMLSchema#dateTime> .
"
`;

exports[`[MarkdownPipeline] "./dot triples.md" 1`] = `
"<http://example.org/dot%20triples.md> <http://example.org/property> \\"value\\" <http://example.org/dot%20triples.md> .
<http://example.org/dot%20triples.md> <http://example.org/same-as> <http://example.org/dot%20triples.md> <http://example.org/dot%20triples.md> .
<http://example.org/subject2> <http://example.org/property2> \\"value2\\" <http://example.org/dot%20triples.md> .
<http://example.org/dot%20triples.md> <http://dottriples.org/path> \\"./dot triples.md\\" .
<http://example.org/dot%20triples.md> <http://dottriples.org/name> \\"dot triples\\" .
<http://example.org/dot%20triples.md> <http://dottriples.org/size> \\"161\\"^^<http://www.w3.org/2001/XMLSchema#integer> .
<http://example.org/dot%20triples.md> <http://dottriples.org/atime> \\"2022-09-25T17:26:24.370Z\\"^^<http://www.w3.org/2001/XMLSchema#dateTime> .
<http://example.org/dot%20triples.md> <http://dottriples.org/mtime> \\"2022-09-25T17:25:21.221Z\\"^^<http://www.w3.org/2001/XMLSchema#dateTime> .
<http://example.org/dot%20triples.md> <http://dottriples.org/ctime> \\"2022-09-25T17:25:21.221Z\\"^^<http://www.w3.org/2001/XMLSchema#dateTime> .
"
`;

exports[`[MarkdownPipeline] "./folder/Test.md" 1`] = `
"<http://example.org/folder/Test.md> <http://example.org/linked-to> <http://example.org/not-found-in-vault> <http://example.org/folder/Test.md> .
<http://example.org/folder/Test.md> <http://dottriples.org/path> \\"./folder/Test.md\\" .
<http://example.org/folder/Test.md> <http://dottriples.org/name> \\"Test\\" .
<http://example.org/folder/Test.md> <http://dottriples.org/size> \\"99\\"^^<http://www.w3.org/2001/XMLSchema#integer> .
<http://example.org/folder/Test.md> <http://dottriples.org/atime> \\"2022-09-25T13:46:50.342Z\\"^^<http://www.w3.org/2001/XMLSchema#dateTime> .
<http://example.org/folder/Test.md> <http://dottriples.org/mtime> \\"2022-08-31T22:05:10.160Z\\"^^<http://www.w3.org/2001/XMLSchema#dateTime> .
<http://example.org/folder/Test.md> <http://dottriples.org/ctime> \\"2022-08-31T22:05:10.160Z\\"^^<http://www.w3.org/2001/XMLSchema#dateTime> .
"
`;

exports[`[MarkdownPipeline] "./index.md" 1`] = `
"<http://example.org/bob/Bob.md> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://example.org/Person.md> <http://example.org/index.md> .
<http://example.org/Alice.md> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://example.org/Person.md> <http://example.org/index.md> .
<http://bob.org> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> \\"WeBsiTe\\" <http://example.org/index.md> .
<http://example.org/index.md> <http://dottriples.org/path> \\"./index.md\\" .
<http://example.org/index.md> <http://dottriples.org/name> \\"index\\" .
<http://example.org/index.md> <http://dottriples.org/size> \\"93\\"^^<http://www.w3.org/2001/XMLSchema#integer> .
<http://example.org/index.md> <http://dottriples.org/atime> \\"2022-09-25T17:26:24.406Z\\"^^<http://www.w3.org/2001/XMLSchema#dateTime> .
<http://example.org/index.md> <http://dottriples.org/mtime> \\"2022-09-25T17:25:33.113Z\\"^^<http://www.w3.org/2001/XMLSchema#dateTime> .
<http://example.org/index.md> <http://dottriples.org/ctime> \\"2022-09-25T17:25:33.113Z\\"^^<http://www.w3.org/2001/XMLSchema#dateTime> .
"
`;
